### ISSUE TITLE
How to create normal .docker/config.json auth file

### ISSUE THREAD

**Issue by ericvicenti** [docker/for-mac#4100]
This was previously reported here, but the issue was auto closed: https://github.com/docker/for-mac/issues/1584

### tldr: On my mac, how can I create a `~/.docker/config.json` file that includes credentials?

The file created on mac has `"credsStore": "osxkeychain"` by default, instead of storing the real credentials. I wonder if there is a way to disable this temporarily.

### Background

If you're curious why I need this file.. I am trying to authenticate my k8s cluster to a private registry. I am trying to run this command:
```
sudo kubectl create secret generic regcred --from-file=.dockerconfigjson=/home/sammy/.docker/config.json --type=kubernetes.io/dockerconfigjson
```
([as instructed by this DigitalOcean kubernetes tutorial](https://www.digitalocean.com/community/tutorials/how-to-set-up-a-private-docker-registry-on-top-of-digitalocean-spaces-and-use-it-with-digitalocean-kubernetes#step-3-%E2%80%94-adding-account-authentication-and-configuring-kubernetes-access))

**Comment by ericvicenti** [comment]
Update: I did not find a workaround on macOS, but I created the file by installing and running `docker login` on linux.

For future reference, you can manually create an `.docker/config.json` authentication file with the following format:

```
{
	"auths": {
		"REGISTRY": {
			"auth": "base64(USERNAME:PASSWORD)"
		}
	},
	"HttpHeaders": {
		"User-Agent": "Docker-Client/18.09.7 (linux)"
	}
}
```

Don't forget to set the correct registry, then run your `username:password` through base64 before putting it into auth. HttpHeaders are probably optional.

**Comment by shivankar-madaan** [comment]
Just sharing my workaround

With Docker-clinet was not able to authenticate via the UI option i.e Sign in
Tried every config file from linux as well. Didn't work

But updated docker and then
Was able to login via UI of Docker desktop service
Here's how my config file looks now
```
{
  "credsStore" : "desktop",
  "HttpHeaders" : {
    "User-Agent" : "Docker-Client/18.06.1-ce (darwin)"
  },
  "auths" : {
    "https://index.docker.io/v1/" : {

    }
  }
}
```

Working well with this

**Comment by leohmoraes** [comment]
Hi, I need an example of config.json with more memory and CPU.
The Docker Toolbox is starting forever and maybe is need do it.

Thanks and sorry for my English

**Comment by abrechon** [comment]
simply remove the "credsStore" block in .docker/config.json and issue docker login. This will disable the helper and you should see credentials in config.json as stated here https://docs.docker.com/engine/reference/commandline/cli/ 

```The property credsStore specifies an external binary to serve as the default credential store. When this property is set, docker login will attempt to store credentials in the binary specified by docker-credential-<value> which is visible on $PATH. If this property is not set, credentials will be stored in the auths property of the config

**Comment by malimccalla** [comment]
@ericvicenti @shivankar-madaan Did either of you manage to find a solution to this? I believe my issue #4545 my be related to this

@abrechon Unfortunately your solution didn't work for me. If I remove the `credsStore` property it only gets added in again after running `docker login`

**Comment by Embraser01** [comment]
@malimccalla for info, I manage to make it work by creating.a small JS script that output a base64 string ready for a k8s secret. Here is the [gist](https://gist.github.com/Embraser01/de9ab4e46f00bdcc725568233ab4cb38)

**Comment by maityh** [comment]
@malimccalla Open the docker preferences and uncheck the option - 'Securely store Docker logins in macOS keychain' then it will not add credsStore property again after running docker login.

**Comment by malimccalla** [comment]
@maityh  This box is not uncheckable for me. It seems to be an ongoing issue https://github.com/docker/for-mac/issues/3774

<img width="657" alt="Screenshot 2020-05-19 at 09 36 17" src="https://user-images.githubusercontent.com/18899991/82304439-3aed1380-99b4-11ea-95ba-85c74378ac55.png">

**Comment by llamahunter** [comment]
FWIW, if someone wants to export the credential store to a temp config.json in a shell script.  Requires 'jq' and 'base64' be installed.

```
    dockercfg=$(mktemp /tmp/dockercfg.XXXXX)
    storetype=$(jq -r .credsStore < ~/.docker/config.json)
    (
        echo '{'
        echo '    "auths": {'
        for registry in $(docker-credential-$storetype list | jq -r 'to_entries[] | .key'); do
            if [ ! -z $FIRST ]; then
                echo '        },'
            fi
            FIRST=true
            credential=$(echo $registry | docker-credential-$storetype get | jq -jr '"\(.Username):\(.Secret)"' | base64)
            echo '        "'$registry'": {'
            echo '            "auth": "'$credential'"'
        done
        echo '        }'
        echo '    }'
        echo '}'
    ) > $dockercfg

    # remove temporary docker config.json when shell script exits
    trap "rm -f $dockercfg" EXIT
```

**Comment by masonlr** [comment]
> I am trying to authenticate my k8s cluster to a private registry.

To add to everything above, you can create local yaml secrets using `--dry-run` with `kubectl`, for example:

```shell
#!/bin/bash

NAME="secret-name"
SERVER="registry.gitlab.com"
USERNAME="test"
PASSWORD="my-password"
DESTINATION="test.yaml"
NAMESPACE="default"

kubectl create secret docker-registry $NAME \
    --dry-run \
    --docker-server=$SERVER \
    --docker-username=$USERNAME \
    --docker-password=$PASSWORD \
    --namespace=$NAMESPACE \
    -o yaml > $DESTINATION  
```


This will create the secret:

```yaml
# test.yaml
apiVersion: v1
data:
  .dockerconfigjson: eyJhdXRocyI6eyJyZWdpc3RyeS5naXRsYWIuY29tIjp7InVzZXJuYW1lIjoidGVzdCIsInBhc3N3b3JkIjoibXktcGFzc3dvcmQiLCJhdXRoIjoiZEdWemREcHRlUzF3WVhOemQyOXlaQT09In19fQ==
kind: Secret
metadata:
  creationTimestamp: null
  name: secret-name
  namespace: default
type: kubernetes.io/dockerconfigjson
```

which can be applied:

```shell
kubectl create -f test.yaml
```

The details can be confirmed using:

```shell
echo eyJhdXRocyI6eyJyZWdpc3RyeS5naXRsYWIuY29tIjp7InVzZXJuYW1lIjoidGVzdCIsInBhc3N3b3JkIjoibXktcGFzc3dvcmQiLCJhdXRoIjoiZEdWemREcHRlUzF3WVhOemQyOXlaQT09In19fQ== | base64 -D
{"auths":{"registry.gitlab.com":{"username":"test","password":"my-password","auth":"dGVzdDpteS1wYXNzd29yZA=="}}}
```

and

```shell
echo dGVzdDpteS1wYXNzd29yZA== | base64 -D
test:my-password
```
